### PR TITLE
Wire CLI args into configuration system

### LINKED ISSUE #1177: Integrate all CLI options into the configuration management system
With #1160, it should be quite simple to add new configurations through the configuration system. We should expose all configurations that are currently configurable via CLI via the configuration files too.

In particular, the tuning options are a good fit for this.

I propose making this work item dependent on #1174: once we have moved the configuration parameters into a class, it will be trivial to drop the class into the the class targeted by PureConfig.

Suggested by @konnov https://github.com/informalsystems/apalache/pull/1160/files#r769408346

### PR DESCRIPTION
**Pull request by shonfeder**
Closes #1177

## Overview

To advance #2013 and friends, we need to make the configuration of various pass executor pipelines

1. easy to reason about, and
2. thread-safe, so that concurrent RPC requests can be handled safely

This work sets us up to address (1) by solving #1177, which further prepares the way for #2073.

## Changes

- The `ApalacheConfig` class has been expanded and reorganized to give a single specification of all the configurable values that determine program behavior, along with their default values.
- The `ConfigManager` class has been cleaned up to improve the loading and overriding logic
- All CLI parsers have been reworked to communicate all their configured values through an instance of `ApalacheConfig`.
  - This involved ensuring that all CLI options are `Option`s, so that the configuration merging logic can differentiate when a value set by the CLI should override something set in one of the other configuration sources.

## Reviewing

The changeset is mostly simpler than the count would imply: many line are just listing on configuration parameters, and about 100 of the removed lines are work that has been deferred to #2074. See my comments on the PR for info.

While reviewing, it's probably simplest to start with the `options.scala` file, followed by the changes to the `ConfigManager.scala`, and then the `ApalacheCommand.scala`: all other changes are correlates of those.

Unfortunately, I think the commits aren't well organized to make review by commit very helpful in this case.

For a high-level description of these changes and where they are headed, see [ADR 22](https://github.com/informalsystems/apalache/blob/cf8cd12f30c708f593ea510037002ac2ed3efbd4/docs/src/adr/022adr-unification-of-configs-and-options.md).

## Followup

- I will be following up this change set with a short ADR documenting the configuration system and what motivated its introduction. See  #2075.

- [x] Tests added for any new code
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [x] Documentation added for any new functionality
- [ ] [Entries added to `./unreleased/`][changelog format] for any new functionality

[changelog format]: https://github.com/informalsystems/apalache/blob/main/CONTRIBUTING.md#how-to-record-a-change

### CODE REVIEW

File: mod-tool/src/main/scala/at/forsyte/apalache/tla/tooling/opt/AbstractCheckerCmd.scala [hunk 1]
```diff
@@ -5,46 +5,55 @@ import org.backuity.clist.{arg, opt}
...
+    val cfg = super.toConfig()
+    cfg.copy(
+        common = cfg.common.copy(file = Some(file)),
+        checker = cfg.checker.copy(config = config, cinit = cinit, init = init, next = next, inv = inv,
+            temporal = temporal, length = length),
+    )
+  }
   override def setCommonOptions(): Unit = {
+    // TODO: rm once OptionProvider is wired in
+    val cfg = configuration.left.map(err => new ConfigurationError(err)).toTry.get
```
Comment (shonfeder) [inline mod-tool/src/main/scala/at/forsyte/apalache/tla/tooling/opt/AbstractCheckerCmd.scala:43]: This is an intermediate step, using the derived configuration to set all values in the option map. In a follow up, we will remove the option map and replace it with a statically verifiable class of options.

File: mod-tool/src/main/scala/at/forsyte/apalache/tla/tooling/opt/ServerCmd.scala [hunk 2]
```diff
@@ -1,17 +1,12 @@
...
-
 import com.typesafe.scalalogging.LazyLogging
 import at.forsyte.apalache.shai
 
 class ServerCmd
     extends ApalacheCommand(name = "server", description = "Run in server mode (in early development)")
     with LazyLogging {
 
-  // Dummy file used for application (esp. output) configuration
-  def file = new File("server")
```
Comment (shonfeder) [inline mod-tool/src/main/scala/at/forsyte/apalache/tla/tooling/opt/ServerCmd.scala:13]: This dummy file is no longer needed now that the `OutputManager` can fall back to the `routine` config value.

File: test/tla/cli-integration-tests.md [hunk 3]
```diff
@@ -3414,7 +3414,7 @@ $ rm -rf ./test-out-dir ./test-run-dir
 ### configuration management: read run-dir from local `.apalache.cfg`
 
 ```sh
-$ echo "run-dir: ./configured-run-dir" > .apalache.cfg
+$ echo "common.run-dir: ./configured-run-dir" > .apalache.cfg
```
Comment (shonfeder) [inline test/tla/cli-integration-tests.md:3417]: Note that this introduces a change to the configuration format, which moves al the previous top-level config values under the `common` key.

File: tla-io/src/main/scala/at/forsyte/apalache/io/ConfigManager.scala [hunk 4]
```diff
@@ -23,25 +24,8 @@ private object Converters {
   implicit val overrideFileReader = ConfigReader.fromString[File](catchReadError(expandedFilePath(_).toFile()))
   // PureConfig's optF will convert None values to appropriate configuration errors
   implicit val featureReader = ConfigReader.fromString[Feature](optF(Feature.fromString))
-}
+  implicit val featureWriter = ConfigWriter.toString[Feature](_.toString)
 
-/**
```
Comment (shonfeder) [inline tla-io/src/main/scala/at/forsyte/apalache/io/ConfigManager.scala:28]: This trait is no longer needed now that every CLI spec is mapped into an instance of `ApalacheConfig`

File: mod-infra/src/main/scala/at/forsyte/apalache/infra/passes/options.scala [hunk 5]
```diff
@@ -64,53 +229,3 @@ object SMTEncoding {
     case oddEncodingType => throw new IllegalArgumentException(s"Unexpected SMT encoding type $oddEncodingType")
   }
 }
-
```
Comment (shonfeder) [inline mod-infra/src/main/scala/at/forsyte/apalache/infra/passes/options.scala:67]: This bits can safely be ignored during review. Followup WIP the implements the options classes can be seen in #2074.

File: tla-io/src/main/scala/at/forsyte/apalache/io/ConfigManager.scala [hunk 6]
```diff
@@ -50,67 +34,74 @@ trait CliConfig {
...
-      // `withFallback` supplies configuration sources that only apply if the preceding configs aren't set
-      .withFallback(globalConfig.optional)
+    val localConfig: ConfigObjectSource =
+      cfg.common.configFile.orElse(findLocalConfig(Paths.get(".").toAbsolutePath())) match {
+        case Some(cfgFile) => ConfigSource.file(cfgFile)
+        case None          => ConfigSource.empty
+      }
+
+    // Derive lightbend `Config` objects from the default and primary config
+    // case classes.
```
Comment (thpani) [inline tla-io/src/main/scala/at/forsyte/apalache/io/ConfigManager.scala:80]: What are the default and primary config case classes?

**Review by thpani — APPROVED** [review]
Phew, that's some heavy lifting 😄 

LGTM!
I'm not too excited about the added logic in the `*Cmd`s' `toConfig` methods, but I also can't think of a better alternative...

**Comment by shonfeder** [comment]
>  I'm not too excited about the added logic in the *Cmds' toConfig methods, but I also can't think of a better alternative...

Ditto! :laughing: 

I take some consolation from thinking that it "only" replaces and expands the logic that had previously been located in the `ConfigManager` class ;)  I also have some hope that if, as per #1624, we end up changing the CLI library and opting for https://github.com/alexarchambault/case-app, we might be able to automate much of that conversion logic away.

Thanks for reviewing!

File: mod-infra/src/main/scala/at/forsyte/apalache/infra/passes/options.scala [hunk 7]
```diff
@@ -1,26 +1,191 @@
...
-    writeIntermediate: Boolean = false,
-    profiling: Boolean = false,
-    features: Seq[Feature] = Seq())
+/**
+ * The basic interface of classes that specify application configurations
+ */
+sealed trait Config[T] {
+
+  /**
+   * Produces a copy of the config group `T` with all it's attributes (attributes nested config groups) set to `None`
```
[the remark this answers is not in the document]
Comment (shonfeder) [reply mod-infra/src/main/scala/at/forsyte/apalache/infra/passes/options.scala:22]: Good point. I added a note to clarify this term a few lines above in 571d14aae0294e415f26bdd0a8e4ae2ea7c6b15e

[the remark this answers is not in the document]
Comment (shonfeder) [reply mod-infra/src/main/scala/at/forsyte/apalache/infra/passes/options.scala]: There were several words missing! Thanks: 7e5b558876d6c182fa0b83a6a0edc6e1e1018775

File: mod-infra/src/main/scala/at/forsyte/apalache/infra/passes/options.scala [hunk 8]
```diff
@@ -1,26 +1,191 @@
...
+   *   A file from which a local configuration is to be read
+   * @param writeIntermediate
+   *   Whether or not to write intermediate data into the `runDir`
+   * @param profiling
+   *   Whether or not to write general profiling data into the `runDir`
+   * @param features
+   *   Enable experimental features
+   */
+  case class Common(
+      routine: Option[String] = Some("UNCONFIGURED-ROUTINE"),
```
[the remark this answers is not in the document]
Comment (shonfeder) [reply mod-infra/src/main/scala/at/forsyte/apalache/infra/passes/options.scala]: That's a good question. I was thinking at the time that this fallback would let us have some useful output from the output manager in absence of a properly configured "routine", but then isn't it just better to raise an error so the routine can be configured correctly? I think so! 56accb5e6f50e599705e7ae4c7008d120ed54067

File: mod-infra/src/main/scala/at/forsyte/apalache/infra/passes/options.scala [hunk 9]
```diff
@@ -1,26 +1,191 @@
...
+    implicit def noneCase[T]: Case.Aux[Option[T], Option[T]] = at(o => o.flatMap(_ => None))
+    // Takes a `Config[T]` to an empty version of the config (with all fields set to `None`)
+    implicit def configCase[T <: Config[T]]: Case.Aux[T, T] = at(cfg => cfg.empty)
+  }
+
+  /**
+   * The common configurations shared by all of Apalache's modes of execution
+   *
+   * @param routine
+   *   The subcommand or process being executed
```
[the remark this answers is not in the document]
Comment (shonfeder) [reply mod-infra/src/main/scala/at/forsyte/apalache/infra/passes/options.scala]: I was thinking this was a more general category, as we could have configured operations that don't correspond to subcommands. E.g., every session in the server could be its own routine, and each RPC call could be a routine, but maybe it is enough to just identify this commands.  It's true that the only use case currently is for commands, and probably best not to overly generalize: e11cf93fcd4b11a8c69ddc2d531b70bf03f0d510

File: mod-infra/src/main/scala/at/forsyte/apalache/infra/passes/options.scala [hunk 10]
```diff
@@ -1,26 +1,191 @@
...
+   * @param file
+   *   The file from which input data can be read
```
[the remark this answers is not in the document]
Comment (shonfeder) [reply mod-infra/src/main/scala/at/forsyte/apalache/infra/passes/options.scala]: Yeah, it's overly general. It came from just picking up the naming convention used in the CLI parameters. I agree `inputfile` is clearer db2ee1e036e270d0d570f273aa7a0f29b49cd0e6

File: tla-io/src/main/scala/at/forsyte/apalache/io/ConfigManager.scala (hunk 6 above)
Comment (shonfeder) [reply tla-io/src/main/scala/at/forsyte/apalache/io/ConfigManager.scala:80]: They are the named variable commented here: 

- The "primary config" is derived from the `cfg` param supplied to this function (see the `@param cfg` in the scaladoc above).
- The default is derived from the default values specified in `ApalacheConfig`, invoked by calling applying the case class.

Comment (thpani) [reply tla-io/src/main/scala/at/forsyte/apalache/io/ConfigManager.scala:80]: Oh, I missed that this refers to the doccomment on `load`. All good!

**Pull request merged**